File: leetcode/three-sum.java
//https://leetcode.com/problems/3sum/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> sums = new ArrayList<>();
        int left = 0;
        int right = 0;
        int currSum = 0;
        
        if(nums.length < 3) return sums;
        
        Arrays.sort(nums);
        
        for(int i = 0; i<nums.length-2; i++){
            if(nums[i] > 0) break;
            
            if(i == 0 || nums[i] != nums[i-1]){
                left = i+1;
                right = nums.length-1;
                while(left<right){
                    currSum = nums[i] + nums[left] + nums[right];
                    if(currSum == 0){
                        sums.add(Arrays.asList(nums[i], nums[left], nums[right]));
                        while(left<right && nums[left] == nums[left+1]) left++;
                        while(left<right && nums[right] == nums[right-1]) right--;
                        left++;
                        right--;
                    }else if(currSum < 0){
                        left++;
                    }else{
                        right--;
                    }
                }
            }
        }
        
        return sums;
    }
}

File: leetcode/three_sum.py
# https://leetcode.com/problems/3sum


class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()

        num_map: dict[int, int] = {}
        sums: list[int] = []

        for i in range(len(nums)):
            num_map[nums[i]] = i

        i_index: int = 0
        j_index: int = 0
        last_i: int = 100001
        last_j: int = 100001

        while i_index < len(nums):
            if nums[i_index] != last_i:
                j_index = i_index + 1
                last_j = 100001
                while j_index < len(nums):
                    if last_j != nums[j_index]:
                        if (
                            num_map.get(0 - nums[i_index] - nums[j_index])
                            and num_map.get(0 - nums[i_index] - nums[j_index]) > j_index
                        ):
                            sums.append([nums[i_index], nums[j_index], 0 - nums[i_index] - nums[j_index]])
                    last_j = nums[j_index]
                    j_index += 1
            last_i = nums[i_index]
            i_index += 1

        return sums
